molvis/molvisbase/include/inviwo/molvisbase/datastructures/molecularstructure.h

```diff
@@ -0,0 +1,145 @@
+/*********************************************************************************
+ *
+ * Inviwo - Interactive Visualization Workshop
+ *
+ * Copyright (c) 2019 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <inviwo/molvisbase/molvisbasemoduledefine.h>
+#include <inviwo/core/common/inviwo.h>
+#include <inviwo/core/datastructures/datatraits.h>
+#include <inviwo/core/util/document.h>
+
+#include <optional>
+
+namespace inviwo {
+
+namespace molvis {
+
+// (structure id, model id, chain id, residue id, atom name, altloc).
+
+struct IVW_MODULE_MOLVISBASE_API Atoms {
+    void updateAtomNumbers();
+    bool empty() const;
+    size_t size() const;
+    void clear();
+
+    std::vector<dvec3> positions;
+    std::vector<double> bfactors;
+    // std::vector<int> structureIds;
+    std::vector<int> modelIds;
+    std::vector<int> chainIds;
+    std::vector<int> residueIds;
+    std::vector<unsigned char> atomNumbers;
+    std::vector<std::string> fullNames;
+
+    std::vector<int> residueIndices;
+};
+
+struct IVW_MODULE_MOLVISBASE_API Residue {
+    size_t id;
+    std::string name;
+    std::string fullName;
+    size_t chainId;
+    std::set<size_t> atoms;
+};
+
+struct IVW_MODULE_MOLVISBASE_API Chain {
+    size_t id;
+    std::string name;
+    std::set<size_t> residues;
+};
+
+using Bond = std::pair<size_t, size_t>;
+
+class IVW_MODULE_MOLVISBASE_API MolecularStructure {
```

Comment:
Does this include all types of molecular structures or is this a special type of structure?
Could add some information on the expectations of this class.
E.g., what are the relationship between atoms, residues, chains and bonds